daemon/hooks/hook_circlog.go

```diff
@@ -0,0 +1,126 @@
+package hooks
+
+import (
+	"fmt"
+	"sync"
+)
+
+const CIRCULARLOG_INIT_SIZE int = 32 << 10
+
+type CircularLog struct {
+	buf         []byte
+	size        int
+	max         int
+	written     int
+	writeCursor int
+	tex         sync.Mutex
+}
+
+func NewCircularLog(max int) (*CircularLog, error) {
+	if max <= 0 {
+		return nil, fmt.Errorf("max must be positive")
+	}
+
+	return &CircularLog{
+		size: CIRCULARLOG_INIT_SIZE,
+		buf:  make([]byte, CIRCULARLOG_INIT_SIZE),
+		max:  max,
+	}, nil
+}
+
+func (cl *CircularLog) Write(data []byte) (int, error) {
+	cl.tex.Lock()
+	defer cl.tex.Unlock()
+	n := len(data)
+
+	// Keep growing the buffer by doubling until
+	// hitting cl.max
+	bufAvail := cl.size - cl.writeCursor
+	// If cl.writeCursor wrapped on the last write,
+	// then the buffer is full, not empty.
+	if cl.writeCursor == 0 && cl.written > 0 {
+		bufAvail = 0
+	}
+	if n > bufAvail && cl.size < cl.max {
+		newSize := cl.size * 2
+		for {
+			if n > newSize-cl.writeCursor && newSize*2 < cl.max {
+				newSize *= 2
+				continue
+			} else if newSize*2 >= cl.max {
+				newSize = cl.max
+				break
+			} else {
+				break
+			}
+		}
+		newBuf := make([]byte, newSize)
+		// Reset write cursor to old size if wrapped
+		if cl.writeCursor == 0 {
+			cl.writeCursor = cl.size
+		}
+		copy(newBuf, cl.buf[:cl.writeCursor])
+		cl.buf = newBuf
+		cl.size = newSize
+	}
+
+	// If data to be written is larger than the max size,
+	// discard all but the last cl.size bytes
+	if n > cl.size {
+		data = data[n-cl.size:]
```

Comment:
Implemented this, but didn't use a `bool` flag, because it wouldn't be concurrency-safe. Another write could sneak in before `Bytes()` is called, and the buffer would no longer be overrun. It would just have wrapped as designed.

Instead of flagging the overrun, I've just altered the first few bytes of the truncated write before I put it into the buffer.

Reply:
Why doesn't `Bytes()` also take the mtx?